File: 0101_symmetric_tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
 
bool isMirror(struct TreeNode* t1, struct TreeNode* t2) {
    if (t1 == NULL && t2 == NULL) {
        return true;
    } else if (t1 == NULL || t2 == NULL) {
        return false;
    } else {
        return (t1->val == t2->val) &&
            isMirror(t1->left, t2->right) &&
            isMirror(t1->right, t2->left);
    }
} 

bool isSymmetric(struct TreeNode* root) {
    return isMirror(root, root);
}

File: 0101/0101_symmetric_tree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type WrapNode struct {
	node  *TreeNode
	level int
}

type Stack struct {
	S []WrapNode
}

func (S *Stack) push(node *TreeNode, level int) {
	w := WrapNode{
		node:  node,
		level: level,
	}
	S.S = append(S.S, w)
}

func (S *Stack) pop() (*TreeNode, int) {
	w := S.S[len(S.S)-1]
	S.S = S.S[:len(S.S)-1]
	return w.node, w.level
}

func (S *Stack) isempty() bool {
	return len(S.S) == 0
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		// EMPTY TREE
		return true
	}

	// BFS TRAVERSAL

	var s Stack
	var result [][]int

	s.push(root, 0)

	for !s.isempty() {
		node, level := s.pop()

		if level >= len(result) {
			result = append(result, []int{})
		}

		if node == nil {
			// I hope test case donot have MaxInt32
			// I just use it represent nil value of int type
			result[level] = append(result[level], math.MaxInt32)
		} else {
			result[level] = append(result[level], node.Val)
			s.push(node.Left, level+1)
			s.push(node.Right, level+1)
		}
	}

	// CHECK
	for i := 0; i < len(result); i++ {
		j := 0
		k := len(result[i]) - 1
		for j < k {
			if result[i][j] != result[i][k] {
				return false
			}
			j++
			k--
		}
	}

	return true
}
